### ISSUE TITLE
Could not resolve type ...

### ISSUE THREAD

**Issue by JerryWang304** [parikhakshat/autoharness#1]
Hi,
The following command failed to generate harness.
```shell
python3 harness.py -L /local/codeql-home/nginx-1.21.1/objs/ -C /local/codeql-home/codeql/ -M 1 -O /local/nginx/autoharness/ -D nginx -G 1 -Y 1 -F "-I /local/codeql-home/nginx-1.21.1/objs -I /local/codeql-home/nginx-1.21.1/src/core -I /local/codeql-home/nginx-1.21.1/src/event -I /local/codeql-home/nginx-1.21.1/src/http -I /local/codeql-home/nginx-1.21.1/src/mail -I /local/codeql-home/nginx-1.21.1/src/misc -I /local/codeql-home/nginx-1.21.1/src/os -I /local/codeql-home/nginx-1.21.1/src/stream -I /local/codeql-home/nginx-1.21.1/src/os/unix" -X ngx_config.h,ngx_core.h
```
Error messages:
```shell
Compiling query plan for /local/codeql-home/codeql/multiargfunc.ql.
ERROR: Could not resolve module cpp. There should probably be a qlpack.yml file declaring dependencies in /local/codeql-home/codeql or an enclosing directory. (/local/codeql-home/codeql/multiargfunc.ql:1,8-11)
ERROR: Could not resolve type Type (/local/codeql-home/codeql/multiargfunc.ql:3,1-5)
ERROR: Could not resolve type Parameter (/local/codeql-home/codeql/multiargfunc.ql:3,30-39)
ERROR: Could not resolve type PointerType (/local/codeql-home/codeql/multiargfunc.ql:6,40-51)
ERROR: Could not resolve type Type (/local/codeql-home/codeql/multiargfunc.ql:9,1-5)
ERROR: Could not resolve type Parameter (/local/codeql-home/codeql/multiargfunc.ql:9,27-36)
ERROR: Could not resolve type PointerType (/local/codeql-home/codeql/multiargfunc.ql:10,65-76)
ERROR: Could not resolve type Function (/local/codeql-home/codeql/multiargfunc.ql:13,6-14)
ERROR: Could not resolve type Type (/local/codeql-home/codeql/multiargfunc.ql:13,18-22)
ERROR: Could not resolve type Parameter (/local/codeql-home/codeql/multiargfunc.ql:14,18-27)
ERROR: Could not resolve type Struct (/local/codeql-home/codeql/multiargfunc.ql:14,91-97)
ERROR: 'result' cannot be used in this context (/local/codeql-home/codeql/multiargfunc.ql:4,3-9)
ERROR: 'result' cannot be used in this context (/local/codeql-home/codeql/multiargfunc.ql:6,3-9)
ERROR: 'result' cannot be used in this context (/local/codeql-home/codeql/multiargfunc.ql:10,3-9)
ERROR: 'result' cannot be used in this context (/local/codeql-home/codeql/multiargfunc.ql:10,47-53)
```

**Comment by parikhakshat** [comment]
This usually happens when you do not install the libraries.
Install the codeql cli tools and the libraries and put them under one folder.

**Comment by parikhakshat** [comment]
Here are the two links:
CLI: https://github.com/github/codeql/releases/tag/codeql-cli%2Fv2.5.6
LIBRARIES: https://github.com/github/codeql/releases/tag/lgtm%2Fv1.27.0 or by git cloning https://github.com/github/codeql

**Comment by 0roman** [comment]
@parikhakshat could you pls give some more details how to setup? I also had problems to follow your documentation. I ran into the same error as @JerryWang304.

I am using pre-compiled cli binaries: https://github.com/github/codeql-cli-binaries/releases/tag/v2.5.6
and libraries: https://github.com/github/codeql/releases/tag/lgtm%2Fv1.27.0 
I use the same versions as described by you.

I put both in the same directory and added codeql to my PATH

Running ```codeql resolve languages ``` gives the following output
```
➜  codeql-home codeql resolve languages                                                                                         
cpp (/home/roman/git/codeql-home/codeql/cpp)
csharp (/home/roman/git/codeql-home/codeql/csharp)
csv (/home/roman/git/codeql-home/codeql/csv)
go (/home/roman/git/codeql-home/codeql/go)
html (/home/roman/git/codeql-home/codeql/html)
java (/home/roman/git/codeql-home/codeql/java)
javascript (/home/roman/git/codeql-home/codeql/javascript)
properties (/home/roman/git/codeql-home/codeql/properties)
python (/home/roman/git/codeql-home/codeql/python)
xml (/home/roman/git/codeql-home/codeql/xml)
```

Running ```codeql resolve qlpacks ``` gives the following output
```
➜  codeql-home codeql resolve qlpacks                                                                                           
codeql-cpp (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/cpp/ql/src)
codeql-cpp-examples (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/cpp/ql/examples)
codeql-cpp-tests (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/cpp/ql/test)
codeql-cpp-upgrades (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/cpp/upgrades)
codeql-csharp (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/csharp/ql/src)
codeql-csharp-examples (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/csharp/ql/examples)
codeql-csharp-tests (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/csharp/ql/test)
codeql-csharp-upgrades (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/csharp/upgrades)
codeql-java (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/java/ql/src)
codeql-java-examples (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/java/ql/examples)
codeql-java-tests (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/java/ql/test)
codeql-java-upgrades (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/java/upgrades)
codeql-javascript (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/javascript/ql/src)
codeql-javascript-examples (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/javascript/ql/examples)
codeql-javascript-tests (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/javascript/ql/test)
codeql-javascript-upgrades (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/javascript/upgrades)
codeql-python (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/python/ql/src)
codeql-python-examples (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/python/ql/examples)
codeql-python-tests (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/python/ql/test)
codeql-python-upgrades (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/python/upgrades)
codeql-suite-helpers (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/misc/suite-helpers)
legacy-libraries-cpp (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/misc/legacy-support/cpp)
legacy-libraries-csharp (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/misc/legacy-support/csharp)
legacy-libraries-java (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/misc/legacy-support/java)
legacy-libraries-javascript (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/misc/legacy-support/javascript)
legacy-libraries-python (/home/roman/git/codeql-home/codeql-lgtm-v1.27.0/misc/legacy-support/python)
legacy-upgrades (/home/roman/git/codeql-home/codeql/legacy-upgrades)
```

I created database for codeql with

```
codeql database create nginx --source-root /home/roman/git/codeql-home/nginx-1.21.0 --language=cpp
```



Running harness.py gives the following error
```
python3 harness.py -L /home/roman/git/codeql-home/nginx-1.21.0/objs/ -C /home/roman/git/codeql-home/codeql/ -M 1 -O /home/roman/git/codeql-home/autoharness/ -D nginx -G 1 -Y 1 -F "-I /home/roman/git/codeql-home/nginx-1.21.0/objs -I /home/roman/git/codeql-home/nginx-1.21.0/src/core -I /home/roman/git/codeql-home/nginx-1.21.0/src/core -I /home/roman/git/codeql-home/nginx-1.21.0/src/http -I /home/roman/git/codeql-home/nginx-1.21.0/src/mail -I /home/roman/git/codeql-home/nginx-1.21.0/src/misc -I /home/roman/git/codeql-home/nginx-1.21.0/src/os -I /home/roman/git/codeql-home/nginx-1.21.0/src/stream -I /home/roman/git/codeql-home/nginx-1.21.0/src/unix" -X ngx_config.h,ngx_core.h                            

Compiling query plan for /home/roman/git/codeql-home/codeql/multiargfunc.ql.
ERROR: Could not resolve module cpp. There should probably be a qlpack.yml file declaring dependencies in /home/roman/git/codeql-home/codeql or an enclosing directory. (/home/roman/git/codeql-home/codeql/multiargfunc.ql:1,8-11)
ERROR: Could not resolve type Type (/home/roman/git/codeql-home/codeql/multiargfunc.ql:3,1-5)
ERROR: Could not resolve type Parameter (/home/roman/git/codeql-home/codeql/multiargfunc.ql:3,30-39)
ERROR: Could not resolve type PointerType (/home/roman/git/codeql-home/codeql/multiargfunc.ql:6,40-51)
ERROR: Could not resolve type Type (/home/roman/git/codeql-home/codeql/multiargfunc.ql:9,1-5)
ERROR: Could not resolve type Parameter (/home/roman/git/codeql-home/codeql/multiargfunc.ql:9,27-36)
ERROR: Could not resolve type PointerType (/home/roman/git/codeql-home/codeql/multiargfunc.ql:10,65-76)
ERROR: Could not resolve type Function (/home/roman/git/codeql-home/codeql/multiargfunc.ql:13,6-14)
ERROR: Could not resolve type Type (/home/roman/git/codeql-home/codeql/multiargfunc.ql:13,18-22)
ERROR: Could not resolve type Parameter (/home/roman/git/codeql-home/codeql/multiargfunc.ql:14,18-27)
ERROR: Could not resolve type Struct (/home/roman/git/codeql-home/codeql/multiargfunc.ql:14,91-97)
ERROR: 'result' cannot be used in this context (/home/roman/git/codeql-home/codeql/multiargfunc.ql:4,3-9)
ERROR: 'result' cannot be used in this context (/home/roman/git/codeql-home/codeql/multiargfunc.ql:6,3-9)
ERROR: 'result' cannot be used in this context (/home/roman/git/codeql-home/codeql/multiargfunc.ql:10,3-9)
ERROR: 'result' cannot be used in this context (/home/roman/git/codeql-home/codeql/multiargfunc.ql:10,47-53)
A fatal error occurred: Could not read BQRS file at /home/roman/git/codeql-home/autoharness/multiarg.bqrs: java.io.FileNotFoundException: /home/roman/git/codeql-home/autoharness/multiarg.bqrs (No such file or directory)
(eventual cause: FileNotFoundException "/home/roman/git/codeql-home/autoharness/multiarg.bqrs (No such file or directory)")
Traceback (most recent call last):
  File "harness.py", line 160, in <module>
    subprocess.check_output("cd "+ args.ql + ";" +args.ql+ "codeql query run multiargfunc.ql -o " + args.output + "multiarg.bqrs -d " + args.ql + args.database +";" + args.ql + "codeql bqrs decode --format=csv " + args.output + "multiarg.bqrs -o " + args.output + "multiarg.csv", shell=True)
  File "/usr/lib/python3.8/subprocess.py", line 415, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'cd /home/roman/git/codeql-home/codeql/;/home/roman/git/codeql-home/codeql/codeql query run multiargfunc.ql -o /home/roman/git/codeql-home/autoharness/multiarg.bqrs -d /home/roman/git/codeql-home/codeql/nginx;/home/roman/git/codeql-home/codeql/codeql bqrs decode --format=csv /home/roman/git/codeql-home/autoharness/multiarg.bqrs -o /home/roman/git/codeql-home/autoharness/multiarg.csv' returned non-zero exit status 2.

```

**Comment by parikhakshat** [comment]
Where is multiargfunc.ql stored? My program should copy the query to the folder that contains both the CLI and Libraries

**Comment by 0roman** [comment]
```multiargfunc.ql``` is only copied to the folder containing the CLI binaries

where is everything stored:
CLI binaries:     ```/home/roman/git/codeql-home/codeql```
codeql Libraries: ```/home/roman/git/codeql-home/codeql-lgtm-v1.27.0``` 
autoharness: ```/home/roman/git/codeql-home/autoharness```
nginx-source-code: ```/home/roman/git/codeql-home/nginx-1.21.0```
nginx-database: ```/home/roman/git/codeql-home/codeql/nginx```

multiargfunc.ql location after executing unsuccessful harness.py
```/home/roman/git/codeql-home/codeql/multiargfunc.ql```
```/home/roman/git/codeql-home/autoharness/multiargfunc.ql```

**Comment by parikhakshat** [comment]
You have to store the CLI binaries and the Codeql libraries under one folder. This can be seen in the error message which shows that it is not able to import the codeql libraries. I will make it more explicit in the documentation. Sorry about that. Another note this program works for any libraries/executables.

**Comment by 0roman** [comment]
@parikhakshat CLI binaries folder and codeql libraries folder were placed both under one folder /home/roman/git/codeql-home. And if you look at my ```codeql resolve packs``` then cpp is resolved correctly, right?

**Comment by parikhakshat** [comment]
codeql resolve packs does not mean that.
It just shows what packs are supported according to the search path.
More information here: https://codeql.github.com/docs/codeql-cli/manual/resolve-qlpacks/.
To make sure all this program works, please put the CODEQL LIBRARIES and CLI binaries under one folder.

**Comment by 0roman** [comment]
okay, thank you @parikhakshat that makes it more clear now. I think it would be great to have a flag to specify the codeql library separately. For some one who has no experience with codeql the documentation is really hard to follow. I was able to solve the issue by copying the the cpp library from inside the  codeql-lgtm-v1.27.0 under /cpp/ql/src/ to the codeql CLI binaries folder:

```
cp -R /home/roman/git/codeql-home/codeql-lgtm-v1.27.0/cpp/ql/src/* /home/roman/git/codeql-home/codeql
```

**Comment by SmllXzBZ** [comment]
@0roman 
Hello, I follow your comment but still have some problems. 
After copying the cpp/ql/src/*, I run harness.py gives the following error
```
python3 harness.py -L /home/test/nginx-1.21.0/objs/ -C /home/test/codeql-h/codeql/ -M 1 -O /home/test/autoharness/ -D nginx -G 1 -Y 1 -F "-I /home/test/nginx-1.21.0/objs -I /home/test/nginx-1.21.0/src/core -I /home/test/nginx-1.21.0/src/event -I /home/test/nginx-1.21.0/src/http -I /home/test/nginx-1.21.0/src/mail -I /home/test/nginx-1.21.0/src/misc -I /home/test/nginx-1.21.0/src/os -I /home/test/nginx-1.21.0/src/stream -I /home/test/nginx-1.21.0/src/os/unix" -X ngx_config.h,ngx_core.h

ERROR: Referenced pack 'codeql/cpp-all' not found. (/home/test/codeql-h/codeql/qlpack.yml:1,1-1)
ERROR: Referenced pack 'codeql/suite-helpers' not found. (/home/test/codeql-h/codeql/qlpack.yml:1,1-1)
A fatal error occurred: Could not resolve library path for /home/test/codeql-h/codeql
A fatal error occurred: Could not read BQRS file at /home/test/autoharness/multiarg.bqrs: java.io.FileNotFoundException: /home/test/autoharness/multiarg.bqrs (No such file or directory)
(eventual cause: FileNotFoundException "/home/test/autoharness/multiarg.bqrs (No such file or directory)")
Traceback (most recent call last):
  File "harness.py", line 160, in <module>
    subprocess.check_output("cd "+ args.ql + ";" +args.ql+ "codeql query run multiargfunc.ql -o " + args.output + "multiarg.bqrs -d " + args.ql + args.database +";" + args.ql + "codeql bqrs decode --format=csv " + args.output + "multiarg.bqrs -o " + args.output + "multiarg.csv", shell=True)
  File "/usr/lib/python3.6/subprocess.py", line 356, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command 'cd /home/test/codeql-h/codeql/;/home/test/codeql-h/codeql/codeql query run multiargfunc.ql -o /home/test/autoharness/multiarg.bqrs -d /home/test/codeql-h/codeql/nginx;/home/test/codeql-h/codeql/codeql bqrs decode --format=csv /home/test/autoharness/multiarg.bqrs -o /home/test/autoharness/multiarg.csv' returned non-zero exit status 2.
```
where is everything stored:
CLI binaries: `/home/test/codeql-h/codeql`
codeql Libraries: `/home/test/codeql-h/codeql-repo`
autoharness: `/home/test/autoharness`
nginx-source-code: `/home/test/nginx-1.21.0`
nginx-database: `/home/test/codeql-h/codeql/nginx`

multiargfunc.ql location after executing unsuccessful harness.py
`/home/test/codeql-h/codeql/multiargfunc.ql`
`/home/test/autoharness/multiargfunc.ql`
